src/software/geom/spline.cpp

```diff
@@ -0,0 +1,91 @@
+#include "software/geom/spline.h"
+
+Spline::Spline(const std::vector<Point>& points) : knots(points)
+{
+    initLinearSegments(points);
+}
+
+Spline::Spline(const std::initializer_list<Point>& points) : knots(points)
+{
+    initLinearSegments(points);
+}
+
+Point Spline::valueAt(double val) const
+{
+    if (val < 0.0 || val > 1.0)
+    {
+        std::stringstream ss;
+        ss << "Tried to evaluate spline at " << val
+           << ", which is outside of domain of the spline: [0,1]";
+        throw std::invalid_argument(ss.str());
+    }
+
+    if (val == 0.0)
+    {
+        return knots.front();
+    }
+
+    if (val == 1.0)
+    {
+        return knots.back();
+    }
```

Comment:
Why the special cases for the start and end?

Reply:
done